Ic dongchen

DongchenZ opened the pull request:
<!--- Did Several changes for IC_process related to NEON datasets -->
<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->

## Description
<!--- Describe your changes in detail -->
1. Small bug fixes for the existing NEON IC_process.
2. Added a new script in inst folder of data.land package that can help generalizing IC process for multi-site settings.
3. updated existing function pft.add.spp and put it under base/db/R folder.
4. added herbaceous calculations and resampling throughout the IC functions.
5. change the way for calculation of total area by automatically detecting subplot area by subplot ID.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [ ] Immediately
- [ ] Within one week
- [x] When possible

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [x] My name is in the list of CITATION.cff
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

mdietze commented on modules/data.land/R/download_NEON_multi_site.R:
```diff
@@ -0,0 +1,115 @@
+    #write into output
+    VEG_INFO[[i]] <- veg_info
+  }
+  fileConn <- file("/projectnb/dietzelab/dongchen/All_NEON_SDA/NEON42/IC/NEON_downloads/log.txt")
```
YOU HAVE TO STOP HARDCODING PATHS!!!!!

mdietze commented on modules/data.land/R/extract_veg.R:
```diff
@@ -70,7 +71,9 @@ extract_veg <- function(new_site, start_date, end_date,
-      obs[obs$species_USDA_symbol != "2PLANT", ] #removes the rows with 2PLANT, this is a NEON specific code that means they could not identify the species 
+      obs <- obs[obs$species_USDA_symbol != "2PLANT" &  
+                   obs$species_USDA_symbol != "2PLANT-H" &
+                   obs$species_USDA_symbol != "CATO6", ] #removes the rows with 2PLANT, this is a NEON specific code that means they could not identify the species 
```
Why would you remove matches to CATO6? That's not a NEON-specific unmatched code, but a specific Hickory species that's fairly common and perfectly valid to include.

DongchenZ replied:
solved

DongchenZ commented on scripts/pft.add.spp.R:
```diff
@@ -14,80 +14,103 @@
-      next
+    ## look up species based on acronyms. (can be either Symbols or IDs)
+    if(ID_flag){
+      species_qry <- glue::glue_sql(paste0("select * from species where \"id\" = '", acro, "'"), .con = con)
```
Answer to the second question: Because I need to confirm that this ID number is correct and it can be referred to a specie record.

DongchenZ replied:
Answer to the first question: good point!

DongchenZ commented on scripts/pft.add.spp.R:
```diff
@@ -14,80 +14,103 @@
+  # default points to psql-pecan.bu.edu.
   if (is.null(con)) {
-    con <- query.bety.con(...)
+    bety <- dplyr::src_postgres(dbname   = "bety",
```
The query.bety.con function does not exsit.

DongchenZ replied:
Already reverted back.

DongchenZ commented on scripts/pft.add.spp.R:
```diff
@@ -14,80 +14,103 @@
+  }
+  
+  #detect if acronym is ID or Symbol
+  if(is.character(acronym[1])){
```
Already fixed this by including a ID argument.

DongchenZ commented on scripts/pft.add.spp.R:
```diff
@@ -14,80 +14,103 @@
 ##' @return Function does not return a value but does print out diagnostic statements.
-##' @author Michael C. Dietze
+##' @author Michael C. Dietze, Dongchen Zhang
 pft.add.spp <- function(pft, acronym, test = TRUE, con = NULL, ...) {
```
the one in the scripts folder should be removed

DongchenZ commented on base/db/R/pft.add.spp.R:
```diff
@@ -0,0 +1,116 @@
+##' @param ... optional arguements for connecting to database (e.g. password, user name, database)
+##' @return Function does not return a value but does print out diagnostic statements.
+##' @author Michael C. Dietze, Dongchen Zhang
+pft.add.spp <- function(pft, acronym, test = TRUE, con = NULL, ...) {
```
not sure how to do this, should I just change the description of the acronym argument?

mdietze commented on base/db/R/pft.add.spp.R:
```diff
@@ -0,0 +1,121 @@
+#-------------------------------------------------------------------------------
```
Git is still detecting this as an entirely new file, not that you moved the file from the scripts directory (it believes you deleted that file, not moved it). In the future PLEASE use git mv to move files, as we've now lost the revision history of this file. Understanding changes is also a lot harder this way as the file appears as if it's 100% new.

DongchenZ replied:
got it!

mdietze commented on base/db/R/pft.add.spp.R:
```diff
@@ -0,0 +1,121 @@
+##' @title Associate species with a PFT.
+##' @param pft String name of the PFT in the database
+##' @param acronym Specie's Symbols.
+##' @param ID Specie's IDs in Bety.
```
1. function documentation is not making it clear that one needs to provide either the acronym or the ID. It's also not making clear where a valid acronym comes from. You probably need to add a @details section
2. `Specie's` is not a word. Please correct to `Species`

mdietze replied:
@DongchenZ please fix

mdietze commented on modules/data.land/inst/doanload_NEON_multi_site.R:
```diff
@@ -0,0 +1,111 @@
+#' Title generalized for download NEON files of multiSites. and automatically detect the most recent available NEON data and if data is problematic, try next year.
```
Fix spelling error in filename

DongchenZ replied:
fixed

mdietze replied:
filename still listed as `doadload_NEON_multi_site.R` please fix via `git mv`

mdietze commented on modules/data.land/inst/doanload_NEON_multi_site.R:
```diff
@@ -0,0 +1,111 @@
+#' @return VEG_INFO: a list object containing veg_info, site IDs of each site.
+#' @export log.txt file describing the progress of download
+
+Download_multiSites_NEON <- function(settings, main_store_dir){
```
generalize name to make it clear that you are downloading VEG data, not that this is a generic NEON download function

DongchenZ replied:
fixed

mdietze replied:
function name still gives now indication that it downloads veg data. please fix.

mdietze commented on modules/data.land/inst/doanload_NEON_multi_site.R:
```diff
@@ -0,0 +1,111 @@
+    #filter on date
+    if(length(which(dataProducts$dataProductCode=="DP1.10098.001")) == 0){
+      print(paste0("no data available for site: ", tempNEON$siteCode, ". jump to the next site!"))
+      log_txt <- c(log_txt, paste0("no data available for site: ", tempNEON$siteCode, ". jump to the next site!"))
```
1. In this approach to creating a log file, the log only gets written if there are no errors anywhere in the loop. Better to write logs incrementally
2. Also better to use the PEcAn.logger package for writing logs

DongchenZ replied:
fixed, but not using PEcAn.logger

mdietze replied:
please update to use PEcAn.logger

mdietze commented on modules/data.land/inst/doanload_NEON_multi_site.R:
```diff
@@ -0,0 +1,111 @@
+    }
+    NEON_dates <- lubridate::date(paste0(dataProducts$availableMonths[[which(dataProducts$dataProductCode=="DP1.10098.001")]], "-01"))
+    closeYear <- unique(lubridate::year(NEON_dates)[which(abs(lubridate::year(NEON_dates)-lubridate::year(settings$state.data.assimilation$start.date)) == min(abs(lubridate::year(NEON_dates)-lubridate::year(settings$state.data.assimilation$start.date))))])
+    target_date <- lubridate::date(paste0(closeYear,"-07-15"))
```
Not sure why target date is being hard-coded rather than being passed as an argument or taken from the settings

DongchenZ replied:
solved

mdietze replied:
not solved, target date is still hard coded

mdietze commented on modules/data.land/inst/doanload_NEON_multi_site.R:
```diff
@@ -0,0 +1,111 @@
+  #starting FOR LOOP
+  for (i in 1:length(settings)) {
+    #match lon and lat
+    tempLat <- as.numeric(settings[[i]]$run$site$lat)
```
If you're working with a multisettings, is it guaranteed that every element in the list will be a settings or do you need to add checks that you're not trying to access the SDA tag or something like that

mdietze commented on modules/data.land/inst/doanload_NEON_multi_site.R:
```diff
@@ -0,0 +1,111 @@
+    
+    #search within 1 month
+    start_date <- ymd(as.Date(NEON_dates[which(daysDiff == min(daysDiff))]))
+    end_date <- start_date %m+% months(1)
```
Definitely should NOT be hardcoding the search window!!

mdietze replied:
As we've discussed verbally, you may need to look +/- multiple YEARS to make sure that you're seeing the full set of available veg plots, and then for each plot you want to find the "best" match in time. It's also been discussed (in an Issue?) that "best" not always be +/- in time, but there are instances (e.g. forward filtering) where you need to only look BACKWARDS in time. I believe that the solution proposed in an earlier meeting was to allow the upper and lower bounds of the search window to be user supplied arguments (and it's OK if both default to 1 mo), that would give the most flexibility for how the search was done.

DongchenZ replied:
It sounds like not a easy fix but rather a redesign of the whole system, so I'd rather have a meeting with you to talk about this!

mdietze replied:
search window still hard coded

mdietze commented on modules/data.land/inst/doanload_NEON_multi_site.R:
```diff
@@ -0,0 +1,111 @@
+    apparentindividual <- neonstore::neon_read(table = "apparentindividual", product = "DP1.10098.001", site = tempNEON$siteCode, start_date = start_date, end_date = end_date, dir = store_dir)
+    
+    #re-download if data is problematic
+    while (is.null(apparentindividual)) {
```
If original search is improved, shouldn't  need this section

mdietze commented on modules/data.land/inst/doanload_NEON_multi_site.R:
```diff
@@ -0,0 +1,111 @@
+    veg_info[[2]] <- filter.date
+    
+    #Set plot size as veg_info[[1]]
+    veg_info[[1]] <- list(area = 400)
```
You discovered this needed to be fixed

mdietze replied:
hard coded plot area still not fixed

mdietze commented on modules/data.land/inst/doanload_NEON_multi_site.R:
```diff
@@ -0,0 +1,111 @@
+    veg_info[[1]] <- list(area = 400)
+    
+    #set site ID
+    veg_info[[3]] <- settings[[i]]$run$site$id
```
Is downstream code expecting this? Would it be better to do something line names(VEG_INFO)[i] <- settings[[i]]$run$site$id

mdietze replied:
Higher level -- given that there's already a function to download NEON for a single site, why is this function duplicating that functionality rather than just calling that function within the for loop?

mdietze commented:
@DongchenZ this PR is building successfully now. If you could address the remaining issues I'm happy to pull in. Note that there's a few things you list as "fixed" in the review comments that haven't yet been updated.

mdietze commented on base/db/R/pft.add.spp.R:
```diff
@@ -0,0 +1,121 @@
+  
+  #detect if we input Symbol or IDs
+  if(!is.null(acronym)){
+    Specie_elements <- acronym
```
please rename as `Species_elements` throughout code

mdietze commented on base/db/R/pft.add.spp.R:
```diff
@@ -0,0 +1,121 @@
+  
+  #initialize bad to store any problematic items.
+  bad_species <- c()
+  bad_pft_speceis <- c()
```
please fix spelling error on species throughout code

mdietze commented on modules/data.land/inst/doanload_NEON_multi_site.R:
```diff
@@ -0,0 +1,111 @@
+    while (is.null(apparentindividual)) {
+      print(paste0("current year: ", closeYear, " is problematic. Try next year"))
+      closeYear <- closeYear + 1
+      target_date <- lubridate::date(paste0(closeYear,"-07-15"))
```
another hard-coded date

mdietze commented on modules/data.land/inst/doanload_NEON_multi_site.R:
```diff
@@ -0,0 +1,111 @@
+      
+      #search within 1 month
+      start_date <- ymd(as.Date(NEON_dates[which(daysDiff == min(daysDiff))]))
+      end_date <- start_date %m+% months(1)
```
another hard coded search window

mdietze reviewed (changes requested):
@DongchenZ could you finish up your old PR so we can get this into the mainline. I don't want to see any more instances of folks on our own team wasting time rediscovering bugs that are sitting in PRs

DongchenZ commented:
> 

Sorry for the late response, I already fixed those bugs long ago but I didn't do the PR so they will probably look the same; I apologize for that, and I already did a new PR that fixed the bugs you mentioned previously.

DongchenZ commented:
Looks like all Github checks are passed! I deleted the old file called: download_NEON_multi_site.R and replaced it with a new script called: Multi_Site_IC_process_script.R and that will help people who is working on creating IC files with multi-site settings. And I hope if you can help me review those changes! Thanks a lot! @mdietze

mdietze commented on modules/data.land/R/cohort2pool.R:
```diff
@@ -19,33 +19,55 @@
+      subplot_ID <- strsplit(unique_subplot[i], " ")[[1]][3]
+      if(subplot_ID %in% c(31, 32, 40, 41)){
+        area <- c(area, 100)
+      }else if(subplot_ID %in% c(21, 23, 39)){
```
This mapping between subplot_ID and plot area seems suspicious to me -- is this relationship between ID and area really universal across all NEON sites? Could you point me to the documentation that confirms this?

mdietze reviewed (commented):
Also didn't catch one spelling error on species (spelled speceis)
